### PR TITLE
Initial upload attachment support

### PR DESCRIPTION
**Pull request by serkan-inci**
Adding support for attachment settings in the .runsettings file.

### CODE REVIEW

**Comment by karanjitsingh** [comment]
> I don't know why these tests have started to fail in the hosted pool, they're working fine on my system. I tried a build on a private agent and things seem to be working https://dev.azure.com/karanjitsingh/jstestadapter/_build/results?buildId=407
> 
> I'll look into it

Fixed

File: src/JSTest/Settings/JSTestSettings.cs [hunk 1]
```diff
@@ -68,6 +68,11 @@ public string JavaScriptTestFrameworkAsString
         [DataMember]
         public string TestFrameworkConfigJson { get; set; }
 
+        [DataMember]
+        public string AttachmentsFolder { get; set; }
+
+        [DataMember]
+        public bool UploadAttachments { get; set; }
```
Comment (acesiddhu) [inline src/JSTest/Settings/JSTestSettings.cs]: i think this should be uploaded in any case. hence we dont need this flag

File: src/JSTest.Runner/Exceptions/Exception.ts [hunk 2]
```diff
@@ -5,13 +5,14 @@ export enum ExceptionType {
     InvalidArgumentsException,
     InvalidMessageException,
     InvalidJSONException,
+    NotFoundException,
```
Comment (acesiddhu) [inline src/JSTest.Runner/Exceptions/Exception.ts]: >NotFoundException [](start = 4, length = 17)

FileNotFoundException?

File: src/JSTest.Runner/ObjectModel/JSTestSettings.ts [hunk 3]
```diff
@@ -30,5 +35,35 @@ export class JSTestSettings {
...
+                throw new Exception('Attachments folder does not exist: ' + attachmentsFolder, ExceptionType.NotFoundException);
+            }
+
+            const lstat = fs.lstatSync(attachmentsFolder);
+            if (!lstat.isDirectory()) {
+                throw new Exception('Attachments folder needs to be a valid directory: ' + attachmentsFolder, ExceptionType.InvalidArgumentsException);
+            }
+
+            // Make sure we have a distinct folder for each test run
+            attachmentsFolder = path.join(attachmentsFolder, `jstestadapter_run_${Date.now()}`);
```
[the remark this answers is not in the document]
Comment (acesiddhu) [reply src/JSTest.Runner/ObjectModel/JSTestSettings.ts]: *process Id

---
In reply to: [253259619](https://github.com/karanjitsingh/jstestadapter/pull/21#discussion_r253259619) [](ancestors = 253259619)

File: src/JSTest.Runner/TestRunner/TestRunner.ts [hunk 4]
```diff
@@ -57,14 +57,31 @@ export class TestRunner {
...
+                let errorMessage: string = null;
+
                 if (message.Version === Constants.messageProtocolVersion) {
-                    this.jsTestSettings = new JSTestSettings(message.Payload);
+                    try {
+                        this.jsTestSettings = new JSTestSettings(message.Payload, this.environment);
+                    }
+                    catch (err) {
+                        error = err;
+                        errorMessage = "TestRunner: Error when reading test settings";
```
Comment (acesiddhu) [inline src/JSTest.Runner/TestRunner/TestRunner.ts]: >when [](start = 58, length = 4)

NIT: while

File: src/JSTest/Settings/JSTestSettings.cs (hunk 1 above)
Comment (serkan-inci) [reply src/JSTest/Settings/JSTestSettings.cs]: Fixed.

File: src/JSTest.Runner/ObjectModel/JSTestSettings.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (serkan-inci) [reply src/JSTest.Runner/ObjectModel/JSTestSettings.ts]: Added process id.

File: src/JSTest.Runner/TestRunner/TestRunner.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (serkan-inci) [reply src/JSTest.Runner/TestRunner/TestRunner.ts]: Changed.

**Pull request merged**